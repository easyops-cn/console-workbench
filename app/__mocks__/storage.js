const storeMap = new Map();

// eslint-disable-next-line no-underscore-dangle
const __reset = () => {
  storeMap.clear();
  storeMap.set('jobs', [
    {
      id: 1,
      name: 'existed test job',
      cmd: 'whoami',
      cwd: '/home/root'
    }
  ]);
};

__reset();

const storage = {
  get: (key, defaults) => (storeMap.has(key) ? storeMap.get(key) : defaults),
  set: (key, value) => {
    storeMap.set(key, value);
  },
  __reset
};

export default storage;
