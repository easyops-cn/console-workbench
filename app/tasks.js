const taskPidSet = new Set();

export const addTaskPid = taskPid => taskPidSet.add(taskPid);
export const removeTaskPid = taskPid => taskPidSet.delete(taskPid);
export const forEachTaskPid = fn => taskPidSet.forEach(fn);
