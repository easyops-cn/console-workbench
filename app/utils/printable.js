import { forEach } from 'lodash';

export default function printable(
  previous = {
    returned: undefined,
    buffer: [],
    cursor: 0
  },
  log = ''
) {
  let { returned, buffer, cursor } = previous;
  let bufferCopy = buffer.slice();
  forEach(log, chr => {
    if (chr === '\x08') {
      if (cursor > 0) {
        cursor -= 1;
      }
    } else {
      bufferCopy[cursor] = chr;
      cursor += 1;
    }
  });
  const separate = bufferCopy.lastIndexOf('\n');
  if (separate !== -1) {
    if (returned === undefined) {
      returned = '';
    } else {
      returned += '\n';
    }
    returned += bufferCopy.slice(0, separate).join('');
    bufferCopy = bufferCopy.slice(separate + 1);
    cursor -= separate + 1;
    // Keep only last 65535 characters
    returned = returned.substr(-65535, 65535);
  }
  return {
    returned,
    buffer: bufferCopy,
    cursor
  };
}
