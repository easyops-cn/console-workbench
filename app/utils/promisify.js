import fs from 'fs';

const promisify = (object, method) => (...args) =>
  new Promise((resolve, reject) => {
    object[method](...args, (err, result) => {
      if (err) {
        reject(err);
      } else {
        resolve(result);
      }
    });
  });

export const fsPromisify = ['readdir', 'stat', 'lstat', 'readlink'].reduce(
  (acc, method) => {
    acc[method] = promisify(fs, method);
    return acc;
  },
  {}
);
