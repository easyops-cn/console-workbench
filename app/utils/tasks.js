import { remote } from 'electron';

const { addTaskPid, removeTaskPid } = remote.require('./tasks');

const taskMap = new Map();

export const findTaskByJobId = jobId => taskMap.get(jobId);

export const addTask = (jobId, task) => {
  addTaskPid(task.pid);
  return taskMap.set(jobId, task);
};

export const removeTaskByJobId = jobId => {
  const task = findTaskByJobId(jobId);
  if (task !== undefined) {
    removeTaskPid(task.pid);
  }
  return taskMap.delete(jobId);
};
