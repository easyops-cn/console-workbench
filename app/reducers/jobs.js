// @flow
import { combineReducers } from 'redux';
import { without } from 'lodash';
import type { Action } from './types';
import {
  ADD_JOB,
  REMOVE_JOB,
  START_JOB,
  STARTED_JOB,
  JOB_OUTPUT,
  STOP_JOB,
  JOB_CLOSE,
  JOB_ERROR,
  JOB_EXIT,
  CLEAR_JOB_OUTPUT,
  ACTIVATE_JOB,
  UPDATE_JOB,
  REPLACE_JOBS
} from '../actions/jobs';
import printable from '../utils/printable';

const ids = (state = [], action: Action) => {
  switch (action.type) {
    case ADD_JOB:
      return [...state, action.job.id];
    case REMOVE_JOB:
      return without(state, action.job.id);
    case REPLACE_JOBS:
      return action.jobs.map(job => job.id);
    default:
      return state;
  }
};

const entities = (state = {}, action: Action) => {
  switch (action.type) {
    case ADD_JOB:
      return {
        ...state,
        [action.job.id]: {
          ...action.job,
          starting: false,
          running: false,
          stopping: false
        }
      };
    case UPDATE_JOB:
      return {
        ...state,
        [action.job.id]: {
          ...state[action.job.id],
          name: action.job.name,
          cmd: action.job.cmd,
          cwd: action.job.cwd,
          subPackageDir: action.job.subPackageDir
        }
      };
    case REMOVE_JOB:
      return {
        ...state,
        [action.job.id]: undefined
      };
    case REPLACE_JOBS:
      return action.jobs.reduce((acc, job) => {
        acc[job.id] = {
          ...job,
          starting: false,
          running: false,
          stopping: false
        };
        return acc;
      }, {});
    case START_JOB:
      return {
        ...state,
        [action.job.id]: {
          ...state[action.job.id],
          starting: true,
          error: undefined
        }
      };
    case STARTED_JOB:
      return {
        ...state,
        [action.job.id]: {
          ...state[action.job.id],
          starting: false,
          running: true
        }
      };
    case JOB_CLOSE:
    case JOB_EXIT:
      if (!state[action.job.id].running) {
        return state;
      }
      return {
        ...state,
        [action.job.id]: {
          ...state[action.job.id],
          running: false,
          stopping: false
        }
      };
    case JOB_ERROR:
      return {
        ...state,
        [action.job.id]: {
          ...state[action.job.id],
          running: false,
          error: action.error
        }
      };
    case STOP_JOB:
      return {
        ...state,
        [action.job.id]: {
          ...state[action.job.id],
          stopping: true
        }
      };
    default:
      return state;
  }
};

const outputs = (state = {}, action) => {
  switch (action.type) {
    case ADD_JOB:
    case REMOVE_JOB:
    case CLEAR_JOB_OUTPUT:
      return {
        ...state,
        [action.job.id]: undefined
      };
    case REPLACE_JOBS:
      return action.jobs.reduce((acc, job) => {
        acc[job.id] = undefined;
        return acc;
      }, {});
    case START_JOB:
      return {
        ...state,
        [action.job.id]: printable(
          undefined,
          `$ cd ${action.job.cwd}\n$ ${action.job.cmd}\n`
        )
      };
    case JOB_OUTPUT:
      return {
        ...state,
        [action.job.id]: printable(state[action.job.id], action.output)
      };
    case JOB_CLOSE:
      return {
        ...state,
        [action.job.id]: printable(
          state[action.job.id],
          `closed with code: ${action.code}, signal: ${action.signal}\n`
        )
      };
    case JOB_EXIT:
      return {
        ...state,
        [action.job.id]: printable(
          state[action.job.id],
          `exited with code: ${action.code}, signal: ${action.signal}\n`
        )
      };
    case STOP_JOB:
      return {
        ...state,
        [action.job.id]: printable(state[action.job.id], '^C\n')
      };
    default:
      return state;
  }
};

const active = (state = null, action) => {
  switch (action.type) {
    case ACTIVATE_JOB:
      return state === action.job.id ? null : action.job.id;
    default:
      return state;
  }
};

const jobs = combineReducers({
  ids,
  entities,
  outputs,
  active
});

export default jobs;
