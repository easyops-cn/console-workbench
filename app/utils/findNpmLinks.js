import path from 'path';
import { sortBy } from 'lodash';
import { fsPromisify } from './promisify';

const readlinkRecursively = async (filePath, depth = 1, MAX_DEPTH = 3) => {
  const target = path.resolve(
    filePath,
    '..',
    await fsPromisify.readlink(filePath)
  );
  if ((await fsPromisify.lstat(target)).isSymbolicLink()) {
    if (depth >= MAX_DEPTH) {
      console.error('max depth for `readlinkRecursively()` reached');
      return target;
    }
    return readlinkRecursively(target, depth + 1, MAX_DEPTH);
  }
  return target;
};

export async function findNpmLinks(dir) {
  const links = [];
  const files = await fsPromisify.readdir(dir, { encoding: 'utf8' });
  await Promise.all(
    files.map(async filename => {
      if (filename.startsWith('.')) {
        return;
      }
      const fileStats = await fsPromisify.lstat(path.join(dir, filename));
      if (fileStats.isSymbolicLink()) {
        links.push({
          packageName: filename,
          link: await readlinkRecursively(path.join(dir, filename))
        });
      } else if (fileStats.isDirectory() && filename.startsWith('@')) {
        const dirWithScope = path.join(dir, filename);
        const filesInScope = await fsPromisify.readdir(dirWithScope, {
          encoding: 'utf8'
        });
        await Promise.all(
          filesInScope.map(async filenameInScope => {
            if (filenameInScope.startsWith('.')) {
              return;
            }
            const fileStatsInScope = await fsPromisify.lstat(
              path.join(dirWithScope, filenameInScope)
            );
            if (fileStatsInScope.isSymbolicLink()) {
              links.push({
                packageName: path.join(filename, filenameInScope),
                link: await readlinkRecursively(
                  path.join(dirWithScope, filenameInScope)
                )
              });
            }
          })
        );
      }
    })
  );
  return sortBy(links, 'packageName');
}
