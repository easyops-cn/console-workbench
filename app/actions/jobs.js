import { spawn } from 'child_process';
import { max } from 'lodash';
import { sync } from 'shell-env';
import defaultShell from 'default-shell';
import { addTask, findTaskByJobId, removeTaskByJobId } from '../utils/tasks';
import storage from '../storage';

export const ADD_JOB = 'ADD_JOB';
export const UPDATE_JOB = 'UPDATE_JOB';
export const REMOVE_JOB = 'REMOVE_JOB';
export const REPLACE_JOBS = 'REPLACE_JOBS';
export const START_JOB = 'START_JOB';
export const STARTED_JOB = 'STARTED_JOB';
export const STOP_JOB = 'STOP_JOB';
export const JOB_OUTPUT = 'JOB_OUTPUT';
export const JOB_CLOSE = 'JOB_CLOSE';
export const JOB_ERROR = 'JOB_ERROR';
export const JOB_EXIT = 'JOB_EXIT';
export const CLEAR_JOB_OUTPUT = 'CLEAR_JOB_OUTPUT';
export const ACTIVATE_JOB = 'ACTIVATE_JOB';

export const addJob = ({ name, cmd, cwd }) => dispatch => {
  const jobs = storage.get('jobs', []);
  const nextId = jobs.length === 0 ? 1 : max(jobs.map(item => item.id)) + 1;
  const newJob = {
    id: nextId,
    name,
    cmd,
    cwd
  };
  storage.set('jobs', [...jobs, newJob]);

  dispatch({
    type: ADD_JOB,
    job: newJob
  });
};

export const updateJob = ({ id, name, cmd, cwd }) => dispatch => {
  const job = { id, name, cmd, cwd };
  const jobs = storage.get('jobs');
  const originalJob = jobs.find(item => item.id === job.id);
  Object.assign(originalJob, job);
  storage.set('jobs', jobs);

  dispatch({
    type: UPDATE_JOB,
    job
  });
};

export const replaceJobs = jobs => dispatch => {
  storage.set('jobs', jobs);

  dispatch({
    type: REPLACE_JOBS,
    jobs
  });
};

export const removeJob = job => dispatch => {
  const jobs = storage.get('jobs');
  storage.set('jobs', jobs.filter(item => item.id !== job.id));

  dispatch({
    type: REMOVE_JOB,
    job
  });
};

const jobOutput = (job, output) => ({
  type: JOB_OUTPUT,
  job,
  output
});

const jobClose = (job, code, signal) => ({
  type: JOB_CLOSE,
  job,
  code,
  signal
});

const jobError = (job, error) => ({
  type: JOB_ERROR,
  job,
  error
});

const jobExit = (job, code, signal) => ({
  type: JOB_EXIT,
  job,
  code,
  signal
});

export const startJob = job => dispatch => {
  if (job.starting || job.running) {
    return;
  }

  dispatch({
    type: START_JOB,
    job
  });

  const task = spawn(job.cmd, [], {
    detached: true,
    cwd: job.cwd,
    env: sync(defaultShell),
    shell: defaultShell
  });

  addTask(job.id, task);

  dispatch({
    type: STARTED_JOB,
    job
  });

  task.stdout.on('data', data => {
    dispatch(jobOutput(job, data.toString()));
  });
  task.stderr.on('data', data => {
    dispatch(jobOutput(job, data.toString()));
  });
  task.on('close', (code, signal) => {
    removeTaskByJobId(job.id);
    dispatch(jobClose(job, code, signal));
  });
  task.on('error', error => {
    removeTaskByJobId(job.id);
    dispatch(jobError(job, error));
  });
  task.on('exit', (code, signal) => {
    removeTaskByJobId(job.id);
    dispatch(jobExit(job, code, signal));
  });
};

export const stopJob = job => dispatch => {
  if (!job.running || job.stopping) {
    return;
  }

  const task = findTaskByJobId(job.id);
  if (task === undefined) {
    return;
  }

  process.kill(-task.pid);

  return dispatch({
    type: STOP_JOB,
    job
  });
};

export const clearJobOutput = job => ({
  type: CLEAR_JOB_OUTPUT,
  job
});

export const activateJob = job => ({
  type: ACTIVATE_JOB,
  job
});
