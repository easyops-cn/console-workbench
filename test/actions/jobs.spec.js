/* eslint-disable import/first */
jest.mock('../../app/storage');

import { spy } from 'sinon';
import * as actions from '../../app/actions/jobs';
import { addTask } from '../../app/utils/tasks';
import storage from '../../app/storage';

describe('jobs actions', () => {
  let originalKill;
  let mockedKill: jest.Mock;

  beforeAll(() => {
    originalKill = process.kill;
    mockedKill = jest.fn();
    Object.defineProperty(process, 'kill', {
      value: mockedKill
    });
    addTask(1, { pid: 65536 });
  });

  beforeEach(() => {
    // eslint-disable-next-line no-underscore-dangle
    storage.__reset();
  });

  it('create addJob action', () => {
    const fn = actions.addJob({
      name: 'test job for addJob action',
      cmd: 'ls -l',
      cwd: '/tmp'
    });
    expect(fn).toBeInstanceOf(Function);

    const dispatch = spy();
    fn(dispatch);

    expect(
      dispatch.calledWith({
        type: actions.ADD_JOB,
        job: {
          id: 2,
          name: 'test job for addJob action',
          cmd: 'ls -l',
          cwd: '/tmp'
        }
      })
    ).toBe(true);
  });

  it('create updateJob action', () => {
    const fn = actions.updateJob({
      id: 1,
      name: 'test job for updateJob action',
      cmd: 'ls -l',
      cwd: '/tmp'
    });
    expect(fn).toBeInstanceOf(Function);

    const dispatch = spy();
    fn(dispatch);

    expect(
      dispatch.calledWith({
        type: actions.UPDATE_JOB,
        job: {
          id: 1,
          name: 'test job for updateJob action',
          cmd: 'ls -l',
          cwd: '/tmp'
        }
      })
    ).toBe(true);
  });

  it('create replaceJobs action', () => {
    const fn = actions.replaceJobs([
      {
        id: 3,
        name: 'test job for replaceJobs action',
        cmd: 'ls -l',
        cwd: '/tmp'
      }
    ]);
    expect(fn).toBeInstanceOf(Function);

    const dispatch = spy();
    fn(dispatch);

    expect(
      dispatch.calledWith({
        type: actions.REPLACE_JOBS,
        jobs: [
          {
            id: 3,
            name: 'test job for replaceJobs action',
            cmd: 'ls -l',
            cwd: '/tmp'
          }
        ]
      })
    ).toBe(true);
  });

  it('create removeJob action', () => {
    const fn = actions.removeJob({
      id: 1,
      name: 'test job for removeJob action',
      cmd: 'ls -l',
      cwd: '/tmp'
    });
    expect(fn).toBeInstanceOf(Function);

    const dispatch = spy();
    fn(dispatch);

    expect(
      dispatch.calledWith({
        type: actions.REMOVE_JOB,
        job: {
          id: 1,
          name: 'test job for removeJob action',
          cmd: 'ls -l',
          cwd: '/tmp'
        }
      })
    ).toBe(true);
  });

  it('create stopJob action', () => {
    const fn = actions.stopJob({
      id: 1,
      name: 'test job',
      running: true
    });
    expect(fn).toBeInstanceOf(Function);

    const dispatch = spy();
    fn(dispatch);

    expect(mockedKill.mock.calls[0][0]).toBe(-65536);

    expect(
      dispatch.calledWith({
        type: actions.STOP_JOB,
        job: {
          id: 1,
          name: 'test job',
          running: true
        }
      })
    ).toBe(true);
  });

  it('create clearJobOutput action', () => {
    expect(
      actions.clearJobOutput({
        id: 1,
        name: 'test job'
      })
    ).toMatchSnapshot();
  });

  it('create activateJob action', () => {
    expect(
      actions.activateJob({
        id: 1,
        name: 'test job'
      })
    ).toMatchSnapshot();
  });

  afterAll(() => {
    Object.defineProperty(process, 'kill', {
      value: originalKill
    });
  });

  /* it('should incrementIfOdd should create increment action', () => {
    const fn = actions.incrementIfOdd();
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    const getState = () => ({ counter: 1 });
    fn(dispatch, getState);
    expect(dispatch.calledWith({ type: actions.INCREMENT_COUNTER })).toBe(true);
  });

  it('should incrementIfOdd shouldnt create increment action if counter is even', () => {
    const fn = actions.incrementIfOdd();
    const dispatch = spy();
    const getState = () => ({ counter: 2 });
    fn(dispatch, getState);
    expect(dispatch.called).toBe(false);
  });

  // There's no nice way to test this at the moment...
  it('should incrementAsync', done => {
    const fn = actions.incrementAsync(1);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    setTimeout(() => {
      expect(dispatch.calledWith({ type: actions.INCREMENT_COUNTER })).toBe(
        true
      );
      done();
    }, 5);
  }); */
});
