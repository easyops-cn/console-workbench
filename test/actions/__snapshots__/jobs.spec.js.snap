// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jobs actions create activateJob action 1`] = `
Object {
  "job": Object {
    "id": 1,
    "name": "test job",
  },
  "type": "ACTIVATE_JOB",
}
`;

exports[`jobs actions create clearJobOutput action 1`] = `
Object {
  "job": Object {
    "id": 1,
    "name": "test job",
  },
  "type": "CLEAR_JOB_OUTPUT",
}
`;
